[DR-3282] Add Pact provider support for integration with WDS consumer

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-3282

This PR does two things:

1. Establishes `datarepo` as a Pact provider, and sets up infrastructure for verifying contracts published against it.
2. Mocks `SnapshotsApiController` dependencies when setting up Pact Provider states used in Workspace Data Service's published contracts against `datarepo`.  Companion PR: https://github.com/DataBiosphere/terra-workspace-data-service/pull/378

**1. General Pact set up**

PRs to `develop` branch will now verify that `datarepo` adheres to the contracts published by its consumers, and will then publish the results to the [central Pact broker](https://pact-broker.dsp-eng-tools.broadinstitute.org/).

I've added instructions for running Pact verification locally, still fetching its Pacts from the central Pact broker but not publishing its results.

It's possible to run this verification on a purely local setup, with a local Pact broker.  It was a bit cumbersome and I didn't get much extra value from it versus verifying Pacts from the broker, so I elected to leave it out of our checked-in documentation.  However, if folks want it codified, I can add it in: https://broadworkbench.atlassian.net/wiki/spaces/IRT/pages/2829680649/Contract+Test+Local+Development

**2. Provider states used in WDS contracts**

This work was driven by the need to test the protected data journey across service boundaries.

WDS is only calling our snapshot retrieval API as part of this journey, so I've spun up a new `SnapshotsApiControllerTest` Pact provider.  It pairs with WDS' published Pacts: https://pact-broker.dsp-eng-tools.broadinstitute.org/pacts/provider/datarepo/consumer/wds/version/0def2c7d6f55390ce00fc8b02e41fe7844ebfa1f

This is presently set up as a pure unit test of `SnapshotsApiController`, so that it can be run quickly with very few prerequisites.  The controller's dependencies are all mocked / stubbed.

**Suggestions for the Pact-curious**

1. Check out this branch and try to get Pact verification working locally for you.  If anything doesn't work or isn't clear, I'd be happy to make changes.
2. Look at the logs associated with new GHA "Verify Consumer Pacts".
3. Explore the [Pact broker](https://pact-broker.dsp-eng-tools.broadinstitute.org/) to see which other services are registered.
4. Join #dsp-contract-testing and read through the [Getting Started with Pact Contract Testing guide](https://broadworkbench.atlassian.net/wiki/spaces/IRT/pages/2660368406/Getting+Started+with+Pact+Contract+Testing).

**Follow-on work**

We've discussed concerns that mocking business logic in our Provider states adds maintenance burden, and increases the possibility that mocks could drift from actual app behavior.  We want to experiment with sharing mocks between our non-Pact unit tests and Pact Provider states.

Alternatively, some other teams here treat their Provider states as [functional tests rather than unit tests](https://broadworkbench.atlassian.net/wiki/spaces/IRT/pages/2498494502/All+Roads+Lead+to+Unit+Tests#Unit-vs-Functional-Checks) -- mocking out external calls and awkward actors but otherwise leaving business logic untouched.  These tests run fast and should not be flaky, but can make it hard to pinpoint where a failure happens.

calypsomatic committed:
first pass at some verification

calypsomatic committed:
test update and gradle task

calypsomatic committed:
some updates

calypsomatic committed:
use more anys

calypsomatic committed:
Create verify_consumer_pacts.yaml

First attempt at copying yaml over

calypsomatic committed:
Update verify_consumer_pacts.yaml

Try a better copy of verify pacts yaml

calypsomatic committed:
rename tdr-provider

okotsopoulos committed:
Show standard stream logging output for Pact verification tests

okotsopoulos committed:
Categorize Pact tests with @Tag

Create reusable Pact.TAG to tag Pact tests, following existing repository convention.

okotsopoulos committed:
Spotless

okotsopoulos committed:
Debug why Pact verification isn't running

okotsopoulos committed:
Set up Pact test with Spring boot test slicing

This follows existing convention in SnapshotsApiControllerTest:
by only loading in the required SnapshotsApiController into the context and mocking all of its inputs,
I want to see what's actually required when setting up the application context before introducing additional env variables.
While here, also broke out the reused UUID into a variable.

okotsopoulos committed:
Try to enable Gradle build scans for Pact tests

And enabled standard logging for easier debugging (default is JSON).

okotsopoulos committed:
Pact broker URL shouldn't specify https prefix

Pact tests were failing with:
au.com.dius.pact.provider.junitsupport.loader.NoPactsFoundException: No Pact files were found to verify
Provider: tdr
Source: Pact Broker http://https://pact-broker.dsp-eng-tools.broadinstitute.org

okotsopoulos committed:
Use HTTPS when resolving Pact Broker

okotsopoulos committed:
Instruct Pact context to use MockMvc as its target

okotsopoulos committed:
GlobalExceptionHandler must be loaded in Pact verification test slicing

Otherwise, when we've configured our mocks to throw, our controller will not be able to properly handle the exceptions.
We've run into this for other MVC unit tests before: see SnapshotsApiControllerTest.java.

okotsopoulos committed:
Pass Pact username and password to Gradle verifyTests target

okotsopoulos committed:
Rename provider to match Helm chart name: datarepo

This will allow for our deployments to be recorded.
This also lays the groundwork to use the same name for both TDR as a consumer and provider, which allows for clearer construction of Pact Broker's network graph.

okotsopoulos committed:
Update documentation to include Pact broker credential management

We needed to make a Terraform change so that the Pact broker credentials could be managed by Atlantis and exposed as GHA env variables.

okotsopoulos committed:
Update au.com.dius.pact.provider:junit5spring to latest

okotsopoulos committed:
Update au.com.dius.pact.provider:junit5 to latest

okotsopoulos committed:
Downgrade au.com.dius.pact.provider to previously working version

Was seeing 'Cannot invoke getPreferredVersion because versionConstraint is null'

calypsomatic commented on src/test/java/bio/terra/pact/provider/DataRepoProviderTest.java:
```diff
@@ -0,0 +1,94 @@
+  @MockBean private AuthenticatedUserRequestFactory authenticatedUserRequestFactory;
+  @MockBean private AssetModelValidator assetModelValidator;
+
+  private static final UUID SNAPSHOT_ID = UUID.fromString("12345678-abc9-012d-3456-e7fab89cd01e");
```
I have yet to learn what the proper procedure is for these dummy variables; here I have defined this on the consumer side and we're repeating the same UUID here, although its exact value doesn't matter.  Is this correct or is there a way to not have to hardcode some specific but unimportant value here?

okotsopoulos replied:
That's a good question!

After reading https://broadworkbench.atlassian.net/wiki/spaces/IRT/pages/2787672065/Demystifying+Provider+States+Callbacks+-+pathFromProviderState and looking at our respective provider and consumer PRs, I am not convinced that matching dummy IDs in this case across provider and consumer is valuable.

WDS as consumer is already defining the desired provider state to work from, so that's what drives the behavior rather than a match of a specific snapshot ID.

I also looked at the Sam provider code and see that their provider states mock off of any IDs rather than specific ones: https://github.com/broadinstitute/sam/blob/5a91c3f81ce8941bb915745bf845782cf1e7d270/pact4s/src/test/scala/org/broadinstitute/dsde/workbench/sam/provider/SamProviderSpec.scala#L150

So I'll change the matchers here from `eq(SNAPSHOT_ID)` to `any(UUID.class)`.  Let me know if that works for you!

okotsopoulos replied:
Actually, I'll flip flop here -- looking at Terra Billing Profile Manager, seems like the hardcoded IDs may be preferable, but I'm still not fully wrapping my head around it, or the best way to share that state between consumers and providers. https://github.com/DataBiosphere/terra-billing-profile-manager/blob/main/service/src/test/java/bio/terra/profile/pact/provider/ProviderStateData.java

calypsomatic replied:
Wow I did not get any notification that you had responded here!  What about the fancy method Josh is using in [these](https://github.com/DataBiosphere/terra-workspace-data-service/pull/372/files) [PRs](https://github.com/DataBiosphere/terra-workspace-manager/pull/1511/files#diff-1363f62b4473f466ae2a9b8865509d5383a815fa61af487b1aa095f7b5ca9031)?

okotsopoulos replied:
@calypsomatic great find, and @jladieu nice work!

I think this should work for us too.  It looks like what we were missing previously was setting up the provider states to accept parameters, like so:

```
  public void authenticatedAsAccount(Map<?, ?> parameters) throws InterruptedException {
    var authenticatedEmail = parameters.get("email").toString();
 ```

I can preemptively make the changes to TDR's snapshot provider states to expect a parameter map with an entry for `id`, and remove our hardcoded, needs-to-match-yours UUID.

okotsopoulos replied:
Done!

okotsopoulos committed:
Simplify Provider states to the minimum required

This reduces ongoing maintenance burden and better reflects the expectations that WDS has of TDR.
Upgraded to latest compatible pact plugins.
Moved away from deprecated host and scheme system properties in favor of setting pactbroker.url.
Since this is getting set in a test profile, we should no longer need to pass it as an environmentn variable in GHA verify_consumer_pacts.

okotsopoulos committed:
Snapshot provider states expect ID to be specified in params

This allowed us to remove a hardcoded dummy snapshot UUID which needed to match the one that WDS was using.
WDS contract tests will need to change to supply this ID.

okotsopoulos committed:
Update Provider state names to reflect expected parameters

These names now match the states referenced by our sole consumer, WDS

jladieu commented on .github/workflows/verify_consumer_pacts.yaml:
```diff
@@ -0,0 +1,205 @@
+          echo "env.CONSUMER_BRANCH=${{ env.CONSUMER_BRANCH }} # This reflects the consumer branch for pact verification (generated by Pact Broker)"
+          echo "env.CONSUMER_SHA=${{ env.CONSUMER_SHA }}       # This reflects the consumer version for pact verification (generated by Pact Broker)"
+          ./gradlew --build-cache verifyPacts --scan
+
```
You might want to add in a task to upload the test reports after verify is concluded.  I found the errors produced without the corresponding report details to be inscrutable in GitHub without it:
```
      - name: Upload Test Reports
        if: always()
        uses: actions/upload-artifact@v1
        with:
          name: Test Reports
          path: service/build/reports
```

Reference [PR](https://github.com/DataBiosphere/terra-workspace-manager/pull/1511/files#diff-beac71d47300076c72974aa63384728f206fd6f49fd016afcba5d51a7d57adecR190-R195)

After this runs with artifact uploads, the test reports with more detailed stacktrace will be available for download from the "Summary" of GHA results page
<img width="245" alt="image" src="https://github.com/DataBiosphere/jade-data-repo/assets/111856/81ad13e8-c41a-4682-b796-11f7307be1d0">

okotsopoulos replied:
Good idea!  But in this case, we are already doing something similar with Gradle scans:

https://scans.gradle.com/s/floparrttw5hu/tests/task/:verifyPacts/details/bio.terra.pact.provider.DataRepoProviderTest?top-execution=1

jladieu commented:
Just dropping a link back to the consumer side of this pact for traceability: https://github.com/DataBiosphere/terra-workspace-data-service/pull/378

okotsopoulos committed:
Remove what I think is a redundant Pact plugin?

okotsopoulos committed:
Add instructions for verifying Pacts locally

We need to populate environment variables with the Pact broker's RO username and password.
This is achieved through an expansion of render-configs.sh, but developers will need to update
however they define their environment variables for this to work.

okotsopoulos committed:
Break out Pact test config into separate script

Only a very small amount of pre-configuration is required to verify Pact contracts: I want a dev to be able to verify contracts without getting bogged down in extraneous set-up.

okotsopoulos committed:
Add consumerVersionSelectors to Pact provider set-up

This makes it so that Pact provider verification will pick up on the expected consumer Pacts from the broker.
Included documentation on how to expand the selector if you wish to pick up Pacts from a consumer's feature branch.
In coordination with WDS consumer, expanded 'user has access to snapshot with given id' provider state to include a table:
this consumer expects that a snapshot retrieved with tables will have at least one table with a name.

okotsopoulos committed:
Rollback SnapshotService method signature changes committed by mistake

okotsopoulos committed:
Add missing SnapshotDaoTest for retrieving nonexistent snapshot

okotsopoulos committed:
Rename DataRepoProviderTest -> SnapshotsApiControllerTest

We want to set boundaries for how we define our Provider states: if future Consumers end up calling other non-snapshot endpoints, we should set up new corresponding Provider state classes.
Note that this name matches existing MockMvc unit test SnapshotsApiControllerTest: I'll seek PR feedback on whether the matching name helps or hurts us.

okotsopoulos committed:
Add unit test to verify SnapshotNotFoundException's class lineage

okotsopoulos committed:
Establish Pact provider state method naming convention

okotsopoulos committed:
Enable test logging for Pact provider verification

okotsopoulos commented on docs/jade-getting-started.md:
```diff
@@ -357,6 +357,10 @@ export AZURE_SYNAPSE_SQLADMINUSER=$(cat /tmp/jade-dev-synapse-admin-user.key)
 export AZURE_SYNAPSE_ENCRIPTIONKEY=$(cat /tmp/jade-dev-synapse-encryption-key.key)
 export AZURE_SYNAPSE_INITIALIZE=false
+
+# Pact contract test settings
```
Note for TDR devs --

These are two new environment variables needed to verify Pacts locally.

okotsopoulos commented on src/test/render-pact-configs.sh:
Note for TDR devs --

I elected to break out this small bit of configuration rendering into its own file. I know that a pain point for new devs onboarding is not being sure which subset of configurations are actually needed to run unit tests, for example.  Interested in your thoughts!

pshapiro4broad replied:
My inclination would be to keep it all in a single file, rather than dividing things up by tests types. A comment in the script can tell a reader which tests require which environment variables.

jladieu commented on src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java:
```diff
@@ -0,0 +1,118 @@
+  @MockBean private AssetModelValidator assetModelValidator;
+
+  @PactBrokerConsumerVersionSelectors
+  public static SelectorBuilder consumerVersionSelectors() {
```
Nice...I'm totally going to copy this.

Do you know what happens when you _don't_ specify this?  Is what's defined here also the default?

okotsopoulos replied:
All credit to @ichengchang who put up a draft PR to model this selection!  I asked that same question in a comment:

https://github.com/DataBiosphere/jade-data-repo/pull/1534#discussion_r1385422359

My most pressing issue at the time was that since TDR had no Pacts published to it from a consumer's **default** branch, TDR Pact verification was reporting no Pacts found.  I could see Pacts published by WDS against TDR, but they weren't being picked up because they were being published by a WDS feature branch.

jladieu reviewed (approved):
Great work.  Also: super thoughtful additions to the readme & shell scripts.

okotsopoulos commented on src/test/java/bio/terra/pact/provider/SnapshotsApiControllerTest.java:
Note for TDR Devs --

You may be thinking, "Don't we already have a `SnapshotsApiControllerTest`?"

We do!  It is a unit test of our `SnapshotsApiController`.  I weirdly liked that these two had the same name (different root directories), it implies a link between the two.  But I am absolutely open to renaming this if preferred.

pshapiro4broad replied:
This seems fine to me, to use the package name to disambiguate tests

pshapiro4broad commented on README.md:
````diff
@@ -82,6 +82,18 @@ When running locally, we are not using the proxy. Therefore, the system doesn't
+To verify that TDR adheres to the contracts published by its consumers, run:
+```
+./src/test/render-pact-configs.sh
+# Reload your environment variables, e.g. src ~/.zshrc
````
So this assumes that the user has added `export PACT_BROKER_USERNAME` etc to their `.zshrc`? I guess overall this doc doesn't really talk about how the rendered configs are exported to the app via environment variables.

pshapiro4broad commented on docs/jade-getting-started.md:
````diff
@@ -372,7 +376,14 @@ export HOST=localhost
+
+Running Pact tests can be achieved by rendering a small set of Pact-specific configurations first:
+```
+./src/test/render-pact-configs.sh
````
I don't see the value in having this script exist separate from render-configs.sh. Is the idea that most of the time we don't run pact tests locally so it's typically not needed?

samanehsan commented:
Will there be any pact tests in TDR for TPS?

okotsopoulos commented:
> Will there be any pact tests in TDR for TPS?

@samanehsan yes!  That work is ticketed separately: https://broadworkbench.atlassian.net/browse/DR-3357

It requires TDR to be set up as a Pact consumer (first time we're doing that) and publish its Pacts against TPS, and for TPS itself to be set up as a Pact provider.  TPS is not yet involved in any Pact testing.

samanehsan reviewed (approved):
Looks great! :octocat:

pshapiro4broad commented on .github/workflows/verify_consumer_pacts.yaml:
```diff
@@ -0,0 +1,205 @@
+# The workflow requires the following Pact broker credentials:
+# - PACT_BROKER_USERNAME - the Pact Broker username
+# - PACT_BROKER_PASSWORD - the Pact Broker password
+# They are managed by Atlantis and were added to Terraform here:
+# https://github.com/broadinstitute/terraform-ap-deployments/pull/1255
```
Up to you but I wouldn't leave this in, in case we change how secrets are populated in the future in GitHub. There's nothing specific to the workflow about this as all secrets in GitHub are populated the same way.

pshapiro4broad commented on .github/workflows/verify_consumer_pacts.yaml:
```diff
@@ -0,0 +1,205 @@
+  pull_request:
+    branches:
+      - develop
+    paths-ignore:
+      - 'README.md'
```
There's nothing wrong with using `paths-ignore` but we don't do it elsewhere in TDR and if you wanted to really do this, there are many other `.md` files which could be listed here.

pshapiro4broad commented on .github/workflows/verify_consumer_pacts.yaml:
```diff
@@ -0,0 +1,205 @@
+      - develop
+    paths-ignore:
+      - 'README.md'
+  push:
```
Why is this run on `push`? That means it will run after a PR is merged to develop, in addition to running before the PR is merged. Under what conditions would this detect an error that would otherwise be missed?

ichengchang replied:
Ultimately, the `verify_cosumer_pacts.yaml` workflow needs to report the `bumper` tag to Pact Broker as the new pact version. The only way to do this will be to publish the verification results once again using the new tag.

pshapiro4broad reviewed (commented):
I see it's already merged, although I have a few minor comments from earlier that I hadn't pushed up yet.
